src/audio/mux/mux.c

```diff
@@ -660,10 +660,63 @@ static int mux_reset(struct comp_dev *dev)
 		}
 	}
 
-	if (dev->ipc_config.type == SOF_COMP_MUX)
-		cd->mux = NULL;
-	else
-		cd->demux = NULL;
+	cd->mux = NULL;
+
+	comp_set_state(dev, COMP_TRIGGER_RESET);
+	return 0;
+}
+
+static int demux_reset(struct comp_dev *dev)
+{
+	int dir = dev->pipeline->source_comp->direction;
+	struct list_item *blist;
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	comp_info(dev, "demux_reset()");
+
+	if (dir == SOF_IPC_STREAM_PLAYBACK) {
+		list_for_item(blist, &dev->bsource_list) {
```

Comment:
demux has only one source, doesn't it? And in fact I don't think we do demux for playback as well as mux for capture, so we could just return errors here and for capture in `mux_reset()` above. Otherwise this introduces some significant code duplication. As an alternative, if you (understandably) don't want to introduce changes to a legacy development branch, we could add static functions with this common code to be called from both `mux_reset()` and `demux_reset()`

Reply:
> demux has only one source, doesn't it? And in fact I don't think we do demux for playback as well as mux for capture, so we could just return errors here and for capture in `mux_reset()` above. Otherwise this introduces some significant code duplication. As an alternative, if you (understandably) don't want to introduce changes to a legacy development branch, we could add static functions with this common code to be called from both `mux_reset()` and `demux_reset()`

If so I would prefer the change of the previous version (before the reset splitting) for code duplication considering, which is:
```
--- a/src/audio/mux/mux.c
+++ b/src/audio/mux/mux.c
@@ -648,12 +647,22 @@ static int mux_reset(struct comp_dev *dev)
                                                                source_list);
                        struct comp_buffer __sparse_cache *sink_c = buffer_acquire(sink);
                        int state = -1;
+                       int direction = -1;
 
-                       if (sink_c->sink)
+                       if (sink_c->sink) {
                                state = sink_c->sink->state;
+                               direction = sink_c->sink->direction;
+                       }
 
                        buffer_release(sink_c);
 
+                       /*
+                        * the sink component is on the opposite direction (PLAYBACK for this case),
+                        * which should be irrelevant to the propagation for reset across the demux.
+                        */
+                       if (direction == SOF_IPC_STREAM_PLAYBACK)
+                               continue;
+
                        if (state > COMP_STATE_READY)
                                /* should not reset the upstream components */
                                return PPL_STATUS_PATH_STOP;
```

because the original code already specified the applicable cases implicitly.
The PPL_STATUS_PATH_STOP checkers under the [if-else clause](https://github.com/thesofproject/sof/blob/rpl-001-drop-stable/src/audio/mux/mux.c#L631) in mux_reset() handle reset propagation in respective cases:
- if dir==PLAYBACK (checking if any active source exists) 
  - for mux (N source to 1 sink): it's possible to meet the checking condition.
  - for demux (1 source to N sink): no way to meet since the propagation flow is from source
- else (dir==CAPTURE; checking if any active sink exists)
  - for mux (N source to 1 sink): no way to meet since the propagation flow is from sink
  - for demux (1 source to N sink): it's possible to meet
 
That is, the original code has the intention for only "mux on playback" and "demux on capture". We don't really need to separate to mux_reset and demux_reset. (I do think adding more comments for explanation would be nice because it is so implicit.)

@lyakh @plbossart thoughts?

Reply:
> I do think adding more comments for explanation would be nice because it is so implicit.

Ok, well, yes, at least your comment above "...to the propagation for reset across the demux" should make it clearer why it only mentions the demux. So, yes, maybe it's better for now to use this smaller change and consider for the future a cleanup, making the limitation of "mux for playback and demux for capture" explicit

Reply:
Updated the commit with more detailed comments.